File: generate.mjs
/**
 * Generate personalized quote-calculator sites from businesses.csv.
 *
 * For each business_name:
 *   - Create generated/<slug>/
 *   - Copy index.html, calculator.js, app.js, styles.css
 *   - Replace {{BUSINESS_NAME}} in HTML + calculator.js
 *
 * Run from the project root:
 *   node generate.mjs
 */

import { readFileSync, writeFileSync, mkdirSync, cpSync, rmSync, existsSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

const ROOT = dirname(fileURLToPath(import.meta.url));
const CSV_PATH = join(ROOT, "businesses.csv");
const OUT_DIR = join(ROOT, "generated");
const TEMPLATE_FILES = ["index.html", "calculator.js", "app.js", "styles.css"];
const TOKEN = "{{BUSINESS_NAME}}";

function slugify(name) {
  const slug = name
    .trim()
    .toLowerCase()
    .replace(/ /g, "-")
    .replace(/[^a-z0-9-]+/g, "")
    .replace(/-{2,}/g, "-")
    .replace(/^-|-$/g, "");
  return slug || "business";
}

/** Minimal CSV parser for a single business_name column (handles quotes). */
function loadBusinessNames(csvPath) {
  if (!existsSync(csvPath)) {
    console.error(
      `Missing businesses.csv. Put your CSV in the project root with a header column named business_name.`
    );
    process.exit(1);
  }

  const raw = readFileSync(csvPath, "utf8").replace(/^\uFEFF/, "");
  const lines = raw.split(/\r?\n/).filter((line) => line.trim() !== "");
  if (lines.length < 2) {
    console.error("No business names found in the CSV.");
    process.exit(1);
  }

  const header = parseCsvLine(lines[0]).map((h) => h.trim());
  const nameIdx = header.indexOf("business_name");
  if (nameIdx === -1) {
    console.error(`CSV must have a header column named business_name (found: ${header.join(", ")})`);
    process.exit(1);
  }

  const names = [];
  for (let i = 1; i < lines.length; i++) {
    const cols = parseCsvLine(lines[i]);
    const name = (cols[nameIdx] || "").trim();
    if (name) names.push(name);
  }

  if (!names.length) {
    console.error("No business names found in the CSV.");
    process.exit(1);
  }

  return names;
}

function parseCsvLine(line) {
  const out = [];
  let cur = "";
  let inQuotes = false;
  for (let i = 0; i < line.length; i++) {
    const ch = line[i];
    if (inQuotes) {
      if (ch === '"') {
        if (line[i + 1] === '"') {
          cur += '"';
          i++;
        } else {
          inQuotes = false;
        }
      } else {
        cur += ch;
      }
    } else if (ch === '"') {
      inQuotes = true;
    } else if (ch === ",") {
      out.push(cur);
      cur = "";
    } else {
      cur += ch;
    }
  }
  out.push(cur);
  return out;
}

function writePersonalizedCopy(businessName, destDir) {
  mkdirSync(destDir, { recursive: true });

  for (const filename of TEMPLATE_FILES) {
    const src = join(ROOT, filename);
    if (!existsSync(src)) {
      console.error(`Missing template file: ${filename}`);
      process.exit(1);
    }

    if (filename === "index.html" || filename === "calculator.js") {
      const text = readFileSync(src, "utf8");
      if (filename === "index.html" && !text.includes(TOKEN)) {
        console.error(`${filename} has no ${TOKEN} token — tokenize the template first.`);
        process.exit(1);
      }
      writeFileSync(join(destDir, filename), text.split(TOKEN).join(businessName), "utf8");
    } else {
      cpSync(src, join(destDir, filename));
    }
  }
}

function main() {
  for (const filename of TEMPLATE_FILES) {
    if (!existsSync(join(ROOT, filename))) {
      console.error(`Missing template file: ${filename}`);
      process.exit(1);
    }
  }

  const names = loadBusinessNames(CSV_PATH);

  if (existsSync(OUT_DIR)) rmSync(OUT_DIR, { recursive: true, force: true });
  mkdirSync(OUT_DIR, { recursive: true });

  const usedSlugs = new Map();
  console.log(`Generating ${names.length} site(s) into generated/ ...`);

  for (const name of names) {
    const base = slugify(name);
    const count = usedSlugs.get(base) || 0;
    usedSlugs.set(base, count + 1);
    const slug = count === 0 ? base : `${base}-${count + 1}`;

    writePersonalizedCopy(name, join(OUT_DIR, slug));
    console.log(`  OK  "${name}" -> generated/${slug}/`);
  }

  console.log("Done.");
}

main();

File: generate.py
"""
Generate personalized quote-calculator sites from businesses.csv.

For each business_name:
  - Create generated/<slug>/
  - Copy index.html, calculator.js, app.js, styles.css
  - Replace {{BUSINESS_NAME}} in HTML + calculator.js

Run from the project root:
  python generate.py
"""

from __future__ import annotations

import csv
import re
import shutil
from pathlib import Path

ROOT = Path(__file__).resolve().parent
CSV_PATH = ROOT / "businesses.csv"
OUT_DIR = ROOT / "generated"
TEMPLATE_FILES = ("index.html", "calculator.js", "app.js", "styles.css")
TOKEN = "{{BUSINESS_NAME}}"


def slugify(name: str) -> str:
    """lowercase, spaces -> dashes, strip other punctuation."""
    text = name.strip().lower()
    text = text.replace(" ", "-")
    text = re.sub(r"[^a-z0-9\-]+", "", text)
    text = re.sub(r"-{2,}", "-", text).strip("-")
    return text or "business"


def load_business_names(csv_path: Path) -> list[str]:
    if not csv_path.is_file():
        raise SystemExit(
            f"Missing {csv_path.name}. Put your CSV in the project root "
            f"with a header column named business_name."
        )

    names: list[str] = []
    with csv_path.open(newline="", encoding="utf-8-sig") as f:
        reader = csv.DictReader(f)
        if not reader.fieldnames or "business_name" not in reader.fieldnames:
            raise SystemExit(
                "CSV must have a header column named business_name "
                f"(found: {reader.fieldnames})"
            )
        for row in reader:
            name = (row.get("business_name") or "").strip()
            if name:
                names.append(name)

    if not names:
        raise SystemExit("No business names found in the CSV.")

    return names


def write_personalized_copy(business_name: str, dest_dir: Path) -> None:
    dest_dir.mkdir(parents=True, exist_ok=True)

    for filename in TEMPLATE_FILES:
        src = ROOT / filename
        if not src.is_file():
            raise SystemExit(f"Missing template file: {filename}")

        if filename in ("index.html", "calculator.js"):
            text = src.read_text(encoding="utf-8")
            if TOKEN not in text and filename == "index.html":
                raise SystemExit(
                    f"{filename} has no {TOKEN} token — tokenize the template first."
                )
            dest_dir.joinpath(filename).write_text(
                text.replace(TOKEN, business_name), encoding="utf-8"
            )
        else:
            shutil.copy2(src, dest_dir / filename)


def main() -> None:
    for filename in TEMPLATE_FILES:
        if not (ROOT / filename).is_file():
            raise SystemExit(f"Missing template file: {filename}")

    names = load_business_names(CSV_PATH)

    if OUT_DIR.exists():
        shutil.rmtree(OUT_DIR)
    OUT_DIR.mkdir(parents=True, exist_ok=True)

    used_slugs: dict[str, int] = {}
    print(f"Generating {len(names)} site(s) into {OUT_DIR}/ ...")

    for name in names:
        base = slugify(name)
        count = used_slugs.get(base, 0)
        used_slugs[base] = count + 1
        slug = base if count == 0 else f"{base}-{count + 1}"

        dest = OUT_DIR / slug
        write_personalized_copy(name, dest)
        print(f"  OK  {name!r} -> generated/{slug}/")

    print("Done.")


if __name__ == "__main__":
    main()
